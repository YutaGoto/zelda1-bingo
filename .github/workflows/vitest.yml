name: vitest

on:
  - push

jobs:
  vitest:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout your repository using git
        uses: actions/checkout@v4
      - name: Setup pnpm
        uses: pnpm/action-setup@v2
        with:
          version: 8
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          cache: "pnpm"
          node-version: 20
      - name: Install dependencies
        run: pnpm install
      - name: Run tests
        run: pnpm test
