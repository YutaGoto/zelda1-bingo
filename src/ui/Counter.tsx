import { AddIcon, MinusIcon } from "@chakra-ui/icons";
import { HStack, IconButton, Input, useNumberInput } from "@chakra-ui/react";

interface CounterProps {
  goal: number;
}

export const Counter = ({ goal }: CounterProps) => {
  const { getInputProps, getIncrementButtonProps, getDecrementButtonProps } =
    useNumberInput({
      step: 1,
      defaultValue: 0,
      min: 0,
      max: goal,
    });

  const inc = getIncrementButtonProps({
    onClick: (e) => {
      e.stopPropagation();
    },
  });
  const dec = getDecrementButtonProps({
    onClick: (e) => {
      e.stopPropagation();
    },
  });
  const input = getInputProps({
    onClick: (e) => {
      e.stopPropagation();
    },
  });

  return (
    <HStack w="95px" marginTop="3">
      <IconButton
        aria-label="decrement"
        size="xs"
        {...dec}
        icon={<MinusIcon />}
      />
      <Input size="xs" {...input} />
      <IconButton
        aria-label="increment"
        size="xs"
        {...inc}
        icon={<AddIcon />}
      />
    </HStack>
  );
};
