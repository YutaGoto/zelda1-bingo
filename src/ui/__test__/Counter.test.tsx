import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { expect, test } from "vitest";

import { Counter } from "../Counter";

const setup = () => {
  const user = userEvent.setup();
  return { user };
};

test("Counter", () => {
  render(<Counter goal={1} marginTop={0} />);

  expect(screen.getByDisplayValue("0")).toBeTruthy();
  expect(screen.getByLabelText("increment")).toBeTruthy();
  expect(screen.getByLabelText("decrement")).toBeTruthy();
});

test("Snapshot", () => {
  const { asFragment } = render(<Counter goal={2} marginTop={0} />);
  expect(asFragment()).toMatchSnapshot();
});

test("Increment", async () => {
  const { user } = setup();
  render(<Counter goal={1} marginTop={0} />);

  await user.click(screen.getByLabelText("increment"));
  expect(screen.getByDisplayValue("1")).toBeTruthy();
});

test("Decrement", async () => {
  const { user } = setup();
  render(<Counter goal={1} marginTop={0} />);

  await user.click(screen.getByLabelText("increment"));
  await user.click(screen.getByLabelText("decrement"));
  expect(screen.getByDisplayValue("0")).toBeTruthy();
});

test("not Increment over goal", async () => {
  const { user } = setup();
  render(<Counter goal={1} marginTop={0} />);

  await user.click(screen.getByLabelText("increment"));
  await user.click(screen.getByLabelText("increment"));
  expect(screen.getByDisplayValue("1")).toBeTruthy();
});
