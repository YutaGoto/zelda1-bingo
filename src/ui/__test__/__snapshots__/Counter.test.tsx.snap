// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="chakra-number-input__root css-wrrj9t"
    data-part="root"
    data-scope="number-input"
    dir="ltr"
    id="number-input:«r1»"
  >
    <div
      class="chakra-stack css-6cpqb8"
    >
      <button
        aria-controls="number-input:«r1»:input"
        aria-label="decrease value"
        class="chakra-button chakra-number-input__decrementTrigger css-u3ytla"
        data-disabled=""
        data-part="decrement-trigger"
        data-scope="number-input"
        dir="ltr"
        disabled=""
        id="number-input:«r1»:dec"
        tabindex="-1"
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
        </svg>
      </button>
      <span
        class="chakra-number-input__valueText css-n812c4"
        data-part="value-text"
        data-scope="number-input"
        dir="ltr"
      >
        0
      </span>
      <button
        aria-controls="number-input:«r1»:input"
        aria-label="increment value"
        class="chakra-button chakra-number-input__incrementTrigger css-u3ytla"
        data-part="increment-trigger"
        data-scope="number-input"
        dir="ltr"
        id="number-input:«r1»:inc"
        tabindex="-1"
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
